src/aspire/source/image.py

```diff
@@ -339,36 +474,69 @@ def set_metadata(self, metadata_fields, values, indices=None):
             values should either be a scalar or a vector of length equal to the total number of images, |self.n|.
         :return: On return, the metadata associated with the specified indices has been modified.
         """
-        # Convert a single metadata field into a list of single metadata field, since that's what the 'columns'
-        # argument of a DataFrame constructor expects.
+
+        # This breaks lots of things, maybe not something we want to rush out.
+        # # Check if we're in an immutable state.
+        # if not self._mutable:
+        #     raise RuntimeError("This source is no longer mutable, try using `update` instead of `set_metadata`")
+
         if isinstance(metadata_fields, str):
             metadata_fields = [metadata_fields]
 
         if indices is None:
-            indices = self._metadata.index.values
+            indices = np.arange(self.n)
 
-        df = pd.DataFrame(values, columns=metadata_fields, index=indices)
-        for metadata_field in metadata_fields:
-            series = df[metadata_field]
-            if metadata_field not in self._metadata.columns:
-                self._metadata = self._metadata.merge(
-                    series, how="left", left_index=True, right_index=True
+        try:
+            iter(values)
+        except TypeError:
+            values = [values] * len(indices)
+        else:
+            if isinstance(values, str):
+                values = [values] * len(indices)
+        assert len(values) == len(
+            indices
+        ), "Mismatch between len(values) and len(indices)"
+
+        values = np.array(values)  # make a copy for our use
+
+        # When creating metadata fields that are string, coerce them into python objects to allow string expansion
+        # later (replacing 'hello' with 'goodbye', for example).
+        # This is in part to conform to legacy implementation of metadata that used pandas,
+        # but is probably a good thing to do anyway.
+        # We also come up with a sensible fill value while we're at it.
+        if np.issubdtype(values.dtype, np.str_):
+            values = values.astype("object")
+            fill_value = ""
+        else:
+            fill_value = np.nan
+
+        if values.ndim == 1:
+            values = values.reshape(-1, 1)  # convert to column
+            values = np.tile(
+                values, (1, len(metadata_fields))
+            )  # stack columns for each metadata field
+
+        for i, metadata_field in enumerate(metadata_fields):
+            if metadata_field not in self._metadata:
+                self._metadata[metadata_field] = np.full(self.n, fill_value).astype(
```

Comment:
Would it make sense to work with `MaskedArray`s here? Maybe something to explore in an issue.

Reply:
Yes, I think so.

Reply:
https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/920